monai/losses/segcalib.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import math
+import warnings
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.modules.loss import _Loss
+
+from monai.utils import pytorch_after
+
+
+def get_gaussian_kernel_2d(ksize: int = 3, sigma: float = 1.0) -> torch.Tensor:
+    x_grid = torch.arange(ksize).repeat(ksize).view(ksize, ksize)
+    y_grid = x_grid.t()
+    xy_grid = torch.stack([x_grid, y_grid], dim=-1).float()
+    mean = (ksize - 1) / 2.0
+    variance = sigma**2.0
+    gaussian_kernel = (1.0 / (2.0 * math.pi * variance + 1e-16)) * torch.exp(
+        -torch.sum((xy_grid - mean) ** 2.0, dim=-1) / (2 * variance + 1e-16)
+    )
+    return gaussian_kernel / torch.sum(gaussian_kernel)
+
+
+class GaussianFilter(torch.nn.Module):
+    def __init__(self, ksize: int = 3, sigma: float = 1.0, channels: int = 0) -> torch.Tensor:
+        super(GaussianFilter, self).__init__()
+        gkernel = get_gaussian_kernel_2d(ksize=ksize, sigma=sigma)
+        neighbors_sum = (1 - gkernel[1, 1]) + 1e-16
+        gkernel[int(ksize / 2), int(ksize / 2)] = neighbors_sum
+        self.svls_kernel = gkernel / neighbors_sum
+        svls_kernel_2d = self.svls_kernel.view(1, 1, ksize, ksize)
+        svls_kernel_2d = svls_kernel_2d.repeat(channels, 1, 1, 1)
+        padding = int(ksize / 2)
+        self.svls_layer = torch.nn.Conv2d(
+            in_channels=channels,
+            out_channels=channels,
+            kernel_size=ksize,
+            groups=channels,
+            bias=False,
+            padding=padding,
+            padding_mode="replicate",
+        )
+        self.svls_layer.weight.data = svls_kernel_2d
+        self.svls_layer.weight.requires_grad = False
+
+    def forward(self, x):
+        return self.svls_layer(x) / self.svls_kernel.sum()
+
+
+class NACLLoss(_Loss):
+    """
+    Murugesan, Balamurali, et al.
```

Comment:
Please add here a brief description of what the loss function does, what applications it's for, and a bit more description of what the arguments do and mean.